File: JavaScript/programmers/무인도여행.js
function solution(maps) {
  const answer = [];
  const visited = Array.from(new Array(maps.length), () =>
    new Array(maps[0].length).fill(0)
  );
  const dx = [0, 0, 1, -1];
  const dy = [1, -1, 0, 0];
  let max = 0;

  for (let i = 0; i < maps.length; i++) {
    for (let j = 0; j < maps[0].length; j++) {
      if (maps[i][j] === "X") {
        visited[i][j] = 1;
      } else {
        if (visited[i][j] === 1) continue;
        visited[i][j] = 1;
        let sum = parseInt(maps[i][j]);
        const queue = [[i, j]];
        while (queue.length) {
          const [x, y] = queue.pop();
          for (let d = 0; d < 4; d++) {
            const nx = x + dx[d];
            const ny = y + dy[d];
            if (nx < 0 || nx >= maps.length || ny < 0 || ny >= maps[0].length)
              continue;
            if (visited[nx][ny] || maps[nx][ny] === "X") continue;
            visited[nx][ny] = 1;
            sum += parseInt(maps[nx][ny]);
            queue.unshift([nx, ny]);
          }
        }
        answer.push(sum);
      }
    }
  }

  return answer.length === 0 ? [-1] : answer.sort((a, b) => a - b);
}

File: javaalgo/programmers/무인도여행.java
package javaalgo.programmers;
import java.util.*;

public class 무인도여행 {
	class Solution {
	    public int[] solution(String[] maps) {
	        ArrayList<Integer> answer = new ArrayList<>();
	        int N = maps.length;
	        int M = maps[0].length();
	        char[][] map = new char[N][M];
	        boolean[][] visited = new boolean[N][M];

	        int[] dx = {0,0,-1,1};
	        int[] dy = {1,-1,0,0};
	        for(int i = 0; i < N; i++){
	            char[] arr = maps[i].toCharArray();
	            for(int j = 0; j < M; j++){
	                map[i][j] = arr[j];
	            }
	        }
	        for(int i = 0; i < N; i++){
	            for(int j = 0; j < M; j++){
	                if(map[i][j] == 'X' || visited[i][j]) continue;
	                Queue<int[]> queue = new LinkedList<>();
	                queue.offer(new int[]{i,j});
	                visited[i][j] = true;
	                int sum = map[i][j] - '0';
	                while(!queue.isEmpty()){
	                    int[] data = queue.poll();
	                    int x = data[0];
	                    int y = data[1];
	                    for(int d = 0; d < 4; d++){
	                        int nx = x + dx[d];
	                        int ny = y + dy[d];
	                        if(nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
	                        if(map[nx][ny] == 'X' || visited[nx][ny]) continue;
	                        sum += map[nx][ny] - '0';
	                        visited[nx][ny] = true;
	                        queue.offer(new int[]{nx,ny});
	                    }
	                }
	                answer.add(sum);
	            }
	        }
	        if(answer.size() == 0) return new int[]{-1};
	        int[] ans = new int[answer.size()];
	        Collections.sort(answer);
	        for(int i =0; i < ans.length; i++){
	            ans[i] = answer.get(i);
	        }
	        return ans;
	    }
	}
}
